File: leetcode153.py
class Solution:
    def findMin(self, nums):
        """
        :type nums: List[int]
        :rtype: int
        """
        if nums[-1]>nums[0]:
            return nums[0]
        ls = len(nums)
        i = ls-1
        while nums[i]>nums[i-1]:
            i-=1
        return nums[i]



class Solution(object):
    def findMin(self, nums):
        """
        type nums: List[int]
        rtype: int
        """
        start, end = 0, len(nums) - 1
        if nums[start] < nums[end]:
                return nums[start]
            
        while start + 1 < end:
            mid = (start + end) // 2
            if nums[start] < nums[end]:
                return nums[start]
            if nums[start] < nums[mid]:
                start = mid
            else:
                end = mid 
                
        return nums[end]

File: leetcodec++/leetcode153.cpp
//
// Created by 黄敏辉 on 2019-08-21.
//

class Solution {
public:
    int findMin(vector<int>& nums) {
        int i = 0, j = nums.size()-1;
        int mid;
        while (i < j){
            if (nums[i] < nums[j])
                return nums[i];

            mid = (i + j) / 2;

            if (nums[mid] >= nums[i])
                i = mid + 1;
            else
                j = mid;
        }
        return nums[i];
    }
};